### PR TITLE
GQ-27 Exceptions/Dev Manager Commentary

### PR DESCRIPTION
**Pull request by popovm**
Requires: #194

### CODE REVIEW

**Comment by popovm** [comment]
Ping @0mega and @bliof for feedback 👀

File: app/models/events/release_exception_event.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+require 'events/base_event'
+
+module Events
+  class ReleaseExceptionEvent < Events::BaseEvent
+    def apps
+      details.fetch('apps', [])
+    end
+
+    def git_repos
+      apps.map { |app| app.fetch('name') }.map { |app_name|
```
Comment (bliof) [inline app/models/events/release_exception_event.rb]: `GitRepositoryLocation.where(name: apps.map { |app| app['name'] }).to_a` ❓

File: app/models/repositories/release_exception_repository.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        comment: event.comment,
+        versions: prepared_versions(event.versions),
+        created_at: event.created_at,
+      )
+    end
+
+    private
+
+    def valid_repo_owner?(event)
+      event.git_repos.all? do |git_repo|
```
Comment (bliof) [inline app/models/repositories/release_exception_repository.rb]: Maybe this should be `any?`. I'll ask in the ticket.

File: features/feature_review.feature [hunk 3]
```diff
@@ -239,9 +240,38 @@ Scenario: QA rejects feature
...
+  Then I should see the Exceptions/Dev Manager Commentary with heading "warning"
+
+  When I "reject" the feature with comment "Not good enough" as a "Repo Owner"
+
+  Then I should see an alert: "Thank you for your submission. It will appear in a moment."
+  When I reload the page after a while
+  Then I should see the Exceptions/Dev Manager Commentary
+    | status | email       | comment         |
+    | danger | foo@bar.com | Not good enough |
+
```
Comment (bliof) [inline features/feature_review.feature]: ✂️

File: features/feature_review.feature [hunk 4]
```diff
@@ -239,9 +240,38 @@ Scenario: QA rejects feature
...
+
+  When I "approve" the feature with comment "This can go live!" as a "Repo Owner"
+
```
Comment (bliof) [inline features/feature_review.feature]: ✂️  Keep at least the When + And together, even When+And+Then

File: spec/models/events/release_exception_event_spec.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+  subject(:event) { described_class.new(details: details) }
+
+  let(:apps) { [{ 'name' => 'frontend', 'version' => 'abc' }] }
+  let(:email) { 'test@example.com' }
+  let(:comment) { 'Critical! Must be released ASAP' }
+  let(:status) { 'approved' }
+
+  let(:default_details) {
+    {
+
```
Comment (bliof) [inline spec/models/events/release_exception_event_spec.rb]: ✂️

File: spec/models/events/release_exception_event_spec.rb [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  describe '#comment' do
+    it 'returns the comment' do
+      expect(event.email).to eq(email)
+    end
+
+    context 'when there is no comment' do
+      let(:details) { default_details.except('comment') }
+
+      it 'returns nil' do
```
Comment (bliof) [inline spec/models/events/release_exception_event_spec.rb]: 🤔  You have to fix this spec..

File: spec/models/repositories/release_exception_repository_spec.rb [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+      t = [4.hours.ago, 3.hours.ago, 2.hours.ago, 1.hour.ago].map { |time| time.change(usec: 0) }
+
+      default = {
+        apps: [%w(app1 1), %w(app2 2)],
+        email: 'test@example.com',
+        comment: 'Good to go',
+        approved: false,
+      }
+
+      events = [
```
Comment (bliof) [inline spec/models/repositories/release_exception_repository_spec.rb]: You could add some method like https://github.com/FundingCircle/shipment_tracker/blob/4ff7d98d3b3f968122cffd192175ec680d3fa360/spec/models/repositories/deploy_repository_spec.rb#L7

e.g. `apply_release_exceptions`

File: app/models/events/release_exception_event.rb [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        GitRepositoryLocation.find_by(name: app_name)
+      }.compact
+    end
+
+    def repo_owner
+      RepoOwner.find_or_create_by(email: email)
```
Comment (popovm) [inline app/models/events/release_exception_event.rb]: The idea was to create an owner for the logged in user, in other words the one who has made the event, and later check if this owner is allowed to create a release exception

File: spec/models/repositories/release_exception_repository_spec.rb (hunk 7 above)
Comment (popovm) [inline spec/models/repositories/release_exception_repository_spec.rb]: You mean `apply` for a collection, don't you?

Comment (bliof) [inline spec/models/repositories/release_exception_repository_spec.rb]: Yep in a way. It should receive multiple hashes and apply them e.g. https://github.com/FundingCircle/shipment_tracker/blob/4ff7d98d3b3f968122cffd192175ec680d3fa360/spec/models/repositories/deploy_repository_spec.rb#L88-L95

In that way you'll remove the array / each / build / apply from here.

**Comment by popovm** [comment]
@bliof and @0mega Could you take a look again? 🐧

**Comment by popovm** [comment]
Hey, @0mega! 
I've made several changes the last two days on the pull request. Could you take another look 👀 
Thanks!

File: spec/models/repositories/release_exception_repository_spec.rb (hunk 7 above)
Comment (bliof) [inline spec/models/repositories/release_exception_repository_spec.rb]: So.. let's extract this into a helper method

File: app/views/feature_reviews/show.html.haml [hunk 9]
```diff
@@ -45,6 +45,10 @@
           %strong
             - icon(item_status_icon_class(@feature_review_with_statuses.uatest_status))
             %span.title User Acceptance Tests
+        %li.list-group-item
+          %strong
+            - icon(item_status_icon_class(@feature_review_with_statuses.release_exception_status))
+            %span.title Exceptions/Dev Maganer Commentary
```
Comment (bliof) [inline app/views/feature_reviews/show.html.haml]: `Repo Owner Commentary` ?

File: app/views/feature_reviews/show.html.haml [hunk 10]
```diff
@@ -130,3 +134,31 @@
           Could not find User Acceptance Tests that ran against expected UAT
           = "(#{@feature_review_with_statuses.uat_url})" if @feature_review_with_statuses.uat_url
           with the app versions that are under review.
+  .col-lg-6
+    - panel(heading: 'Exceptions/Dev Manager Commentary', klass: 'release-exception', status: @feature_review_with_statuses.release_exception_status) do
```
Comment (bliof) [inline app/views/feature_reviews/show.html.haml]: 👁

File: app/views/feature_reviews/show.html.haml [hunk 11]
```diff
@@ -130,3 +134,31 @@
...
+      .panel-body
+        - if @feature_review_with_statuses.release_exception
+          %p
+            %span.repo-owner-email
+              %strong= @feature_review_with_statuses.release_exception.repo_owner.email
+            at
+            %span.repo-owner-time= @feature_review_with_statuses.release_exception.created_at
+          = simple_format(@feature_review_with_statuses.release_exception.comment, class: 'repo-owner-comment')
+        - else
+          No unapproved deployment exception explanation by a Dev Manager
```
Comment (bliof) [inline app/views/feature_reviews/show.html.haml]: 👁

File: app/views/feature_reviews/show.html.haml [hunk 12]
```diff
@@ -130,3 +134,31 @@
           Could not find User Acceptance Tests that ran against expected UAT
           = "(#{@feature_review_with_statuses.uat_url})" if @feature_review_with_statuses.uat_url
           with the app versions that are under review.
+  .col-lg-6
+    - panel(heading: 'Exceptions/Dev Manager Commentary', klass: 'release-exception', status: @feature_review_with_statuses.release_exception_status) do
+      .panel-body
+        - if @feature_review_with_statuses.release_exception
+          %p
+            %span.repo-owner-email
+              %strong= @feature_review_with_statuses.release_exception.repo_owner.email
```
Comment (bliof) [inline app/views/feature_reviews/show.html.haml]: You can use the format_owner_emails helper

File: app/models/release_exception.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+class ReleaseException
+  include Virtus.value_object
+
+  values do
+    attribute :repo_owner_id, Integer
+    attribute :comment, String
+    attribute :approved, Boolean
+    attribute :created_at, Time
```
Comment (bliof) [inline app/models/release_exception.rb]: maybe it is worth adding `#approved_at` which will return the `#created_at` if the release exception is approved  and `#declined_at` when it is unapproved

File: db/migrate/20170105131624_create_release_exceptions.rb [hunk 14]
```diff
@@ -0,0 +1,11 @@
+class CreateReleaseExceptions < ActiveRecord::Migration
+  def change
+    create_table :release_exceptions do |t|
+      t.references :repo_owner, index: true, foreign_key: true
+      t.text :versions, array: true, default: []
+      t.boolean :approved
+      t.text :comment
+      t.timestamps
```
Comment (bliof) [inline db/migrate/20170105131624_create_release_exceptions.rb]: `null: false` and most probably there should be an index on `created_at`

File: features/support/scenario_context.rb [hunk 15]
```diff
@@ -27,7 +27,7 @@ def initialize(app, host)
       @repository_locations = {}
     end
 
-    def setup_application(name)
+    def setup_application(name, repo_owner_email = nil)
```
Comment (bliof) [inline features/support/scenario_context.rb]: let's change it to a hash so you can use it like this:

```
setup_application('my-test-app', owners: 'john@wall.com, ninja@hello.asdf')
```

**Comment by popovm** [comment]
@bliof Could take a look again 👀

**Pull request merged**